iset.mm

```diff
@@ -97612,6 +97825,32 @@ equinumerous to omega (the set of finite ordinal numbers).  (Contributed
   nnct $p |- NN ~<_ _om $=
     ( cn com cen wbr cdom nnenom endom ax-mp ) ABCDABEDFABGH $.
 
+  ${
+    $d i n $.
+    fxnn0nninf.g $e |- G = frec ( ( x e. ZZ |-> ( x + 1 ) ) , 0 ) $.
+    fxnn0nninf.f $e |- F = ( n e. _om |-> ( i e. _om
+      |-> if ( i e. n , 1o , (/) ) ) ) $.
+    $( A function from ` NN0 ` into ` NN+oo ` .  (Contributed by Jim Kingdon,
+       16-Jul-2022.) $)
+    fnn0nninf $p |- ( F o. `' G ) : NN0 --> NN+oo $=
+      ( com xnninf wf cn0 ccnv ccom wel c1o c0 cif cmpt cv wf1o fmpti f1of mp2b
+      nnnninf frechashgf1o f1ocnv fco mp2an ) HIDJKHELZJZKIDUIMJCHIBHBCNOPQRDGB
+      CSUDUAHKETKHUITUJAEFUEHKEUFKHUIUBUCKHIDUIUGUH $.
+
+    fxnn0nninf.i $e |- I = ( ( F o. `' G )
+      u. { <. +oo , ( _om X. { 1o } ) >. } ) $.
+    $( A function from ` NN0* ` into ` NN+oo ` .  (Contributed by Jim Kingdon,
+       16-Jul-2022.) $)
+    fxnn0nninf $p |- I : NN0* --> NN+oo $=
+      ( cxnn0 xnninf wf cn0 cpnf csn cun com c1o wceq wcel ccnv ccom cxp cop wa
+      cin c0 fnn0nninf wf1o pnfex omex 1oex snex xpex f1osn f1of ax-mp infnninf
+      wss snssi mp2an pm3.2i cv wn disj nn0nepnf neneqd elsni nsyl mprgbir fun2
+      fss feq1i mpbir df-xnn0 feq2i ) JKFLMNOZPZKFLZVSVRKDEUAUBZNQROZUCZUDOZPZL
+      ZMKVTLZVQKWCLZUEMVQUFUGSZWEWFWGABCDEGHUHVQWBOZWCLZWIKUSZWGVQWIWCUIWJNWBUJ
+      QWAUKRULUMUNUOVQWIWCUPUQWBKTWKURWBKUTUQVQWIKWCVLVAVBWHAVCZVQTZVDAMAMVQVEW
+      LMTZWLNSWMWNWLNWLVFVGWLNVHVIVJMVQKVTWCVKVAVRKFWDIVMVNJVRKFVOVPVN $.
+  $}
+
```

Comment:
I still wonder how the elements of NN+oo look like. Maybe an example will help, e.g., what **is** `` ( I ` 1 ) ``.
But what is most interesting: what **is** `` ( I ` 0 ) `` (the element of NN+oo corresponding to 0). If NN+oo has such a "0", the it would be better to name it NN0+oo (regardless how it is represented in HTML), wouldn't it?

Reply:
Maybe add to the comment:
> One checks (see also comments of ~infnninf and ~nnnninf) that the function I maps the natural number i to the sequence which "becomes zero" at index i, and maps infinity to the sequence which "never becomes zero", that is, the constant sequence equal to 1.

One would need to prove, with same $e-hypotheses:
```
inftonninf $p |- ( N e. _om -> ( I ` N ) = if ( i e. N , 1o , (/) ) ) $=
nntonninf $p |- ( I ` +oo ) = ( _om X. { 1o } ) $=
```
(but I  haven't looked at the detail of your proofs... This looks redundant with infnninf and nnnninf, but maybe this appears in some subproofs too.

Reply:
As for the token, see discussion at #2621. Simply stated,NN0+oo would be a bit heavy. (The original sin is using NN0 for the natural numbers in my opinion.)

Reply:
I've added comments to infnninf and nnnninf to clarify what elements of `NN+oo` look like. Let us know @avekens if the revised wording there makes it clearer.

It is a bit unclear at least in my mind what we'll state in terms of `I` explicitly and what we'll state in other ways (for example `NN0* ~<_ NN+oo` or `_om e. Omni <-> NN0* ~~ NN+oo`). But that's mostly for future pull requests.

As for the notation, I agree with @benjub that trying to put `0` into the name, either in the source or in the typeset version, is not a good idea. I guess what made me come around to this is partly that `NN+oo` is, while related to natural numbers (as seen in terminology like "one-point compactification of the natural numbers"), its own thing (defined as a sequence).

Reply:
The statements `NN0* ~<_ NN+oo` and `_om e. Omni <-> NN0* ~~ NN+oo` should be proved as corollaries of
```
I : NN0* -1-1-->  NN+oo
_om e. Omni <-> I : NN0* -1-1-onto-> NN+oo
```
since the latter give more information. This is similar to the fact that knowing that two things are isomorphic is nice, but it's nicer to know that a given map is an isomorphism. You can see this in the current trends in (homotopy) type theory.

As an example, you may know Whitehead's theorem: if a map between CW complexes induces isomorphisms on all homotopy groups, then that map is a homotopy equivalence. However, two CW complexes with isomorphic homotopy groups need not be homotopy equivalent.

Reply:
The new comments of infnninf and nnnninf are very helpful. Two add two additional theorems like the following would make things even clearer:

``0nninf $p |- ( I ` 0 ) = ( x e. _om |-> (/) )``
``1nninf $p |- ( I ` 1 ) = ( x e. _om |-> if ( x = (/) , 1o , (/) ) )``

As for the naming, I am afraid that `NN+oo` could be confused with `NN* := ( NN u. { +oo } )` (not defined yet) - that was at least my first expectation.

Reply:
I have added ``( I ` 0 ) = ( x e. _om |-> (/) )`` and ``( I ` 1 ) = ( x e. _om |-> if ( x = (/) , 1o , (/) ) )`` . Hopefully it isn't too much scope creep to add it to this pull request (if there is some controversial thing there, I suppose I should make it a separate branch so we can get this one merged).

Reply:
These examples maybe useful because in theses two cases (especially for 0), the formula simplifies, but more useful would be the above two:
```
inftonninf $p |- ( N e. _om -> ( I ` N ) = ( n e. _om |-> if ( n e. N , 1o , (/) ) ) ) $=
nntonninf $p |- ( I ` +oo ) = ( _om X. { 1o } ) $=
```
(for the second, either using direct product or function notation as you did for 0nninf).

Reply:
> $p |- ( I ` +oo ) = ( _om X. { 1o } ) $=

Already present as
```
    inftonninf $p |- ( I ` +oo ) = ( x e. _om |-> 1o ) $=
```

> $p |- ( N e. _om -> ( I ` N ) = ( n e. _om |-> if ( n e. N , 1o , (/) ) ) ) $=

I have no objection, so go for it.

Reply:
In the comment of `df-nninf $a |- NN+oo = { f e. ( 2o ^m _om ) | ...`, you write:
> This definition would function much the same whether we used ` _om ` or ` NN0 ` , but the former allows us to take advantage of ` 2o = { (/) , 1o } ` ( ~ df2o3 ) so we adopt it.

With all the proofs using the injection ` I ` which is defined using the frightening `frec`, is this still true that `_om` is more convenient ? Sorry I didn't catch this earlier, but given the proximity of NN+oo with NN0, it seems strange to resort to using `( 2o ^m _om )` instead of `( { 0 , 1 } ^m NN0 )` ?

Reply:
Another belated comment:
```
nnnninf $p |- ( N e. _om -> ( i e. _om |-> if ( i e. N , 1o , (/) ) ) e. NN+oo ) $=
```
This is also true if `N = _om`. I don't know how you write it, especially for the intuitionistic version:
```
|- ( N e. succ _om ->
|- ( N e. ( _om u. { _om } )->
|- ( ( N e. _om \/ N = _om ) ->
```
?

Reply:
> With all the proofs using the injection `I` which is defined using the frightening `frec`,

I guess I'm just so used to converting between `_om` and `NN0` that I don't find the particular `frec((𝑥 ∈ ℤ ↦ (𝑥 + 1)), 0)` idiom as frightening as `frec` in general. Cumbersome or clutter, yes. But it is well behaved in light of https://us.metamath.org/ileuni/frec2uzf1od.html and other similar theorems.

> is this still true that `_om` is more convenient ? Sorry I didn't catch this earlier, but given the proximity of NN+oo with NN0, it seems strange to resort to using `( 2o ^m _om )` instead of `( { 0 , 1 } ^m NN0 )` ?

I guess I'm just so used to the literature which pretty much universally uses notation like 2<sup>ℕ</sup> and maybe I'm taking that too literally. I could probably be talked into `{ 0 , 1 }` (especially if I don' t have to do all the work of switching).

Reply:
> I don't know how you write it,

In https://us.metamath.org/ileuni/df-ihash.html we use `( _om u. { _om } )` but all three of the ones you write should be equivalent even intutionistically.

Of course this trick of using `e.` to mean both "element of the the natural numbers" or "natural number less than a given natural number" only works if we use `_om` rather than `NN0` (which I'm not sure is a good reason to use `_om` but is worth pointing out).

Reply:
As for the extension of nnnninf, I added it as a remark in the third commit of #2705. I wrote that it generalizes to "N e. succ _om", even if "succ _om" is not very natural in this context.  If using NN0 in place of _om, this would read "N e. NN0x".